# **The neuron — weights, bias, and one forward pass**

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-ML-Explainers/main/assets/rainbow-divider.png)

# 📖 TABLE OF CONTENTS

- [Setup](#Setup)
- [A neuron is embarrassingly simple](#A-neuron-is-embarrassingly-simple)
  - [Compute one neuron by hand](#Compute-one-neuron-by-hand)
  - [The bias moves the answer alone](#The-bias-moves-the-answer-alone)
  - [The non-linearity that makes it deep](#The-non-linearity-that-makes-it-deep)
- [References](#References)

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-ML-Explainers/main/assets/rainbow-divider.png)

# Setup

Before we jump in, let's quickly run the setup cell. It checks the environment (GPU, library versions) and sets random seeds for reproducibility — we want identical results on every run.

In [1]:
# 🔧 Setup: Run this cell first!
# Environment check + random seeds for reproducibility

import sys
import time
import random
import numpy as np
import pandas as pd
import matplotlib as mpl
import torch
import torch.nn as nn

print(f"📦 Python  {sys.version.split()[0]}")
print(f"🔢 NumPy   {np.__version__}")
print(f"📊 Matplotlib   {mpl.__version__}")
print(f"🔥 PyTorch {torch.__version__}")

# GPU check (only needed for optional deep-feature experiments)
if torch.cuda.is_available():
    device = torch.device("cuda")
    print(f"✅ GPU available: {torch.cuda.get_device_name(0)}")
    print(f"   Memory: {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB")
else:
    device = torch.device("cpu")
    print("⚠️ No GPU detected")

DEVICE_LABEL = "GPU" if device.type == "cuda" else "CPU"

# Random seeds for reproducibility
SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed_all(SEED)

print(f"\n🎲 Random seed set to {SEED} (Python, NumPy, PyTorch)")

%matplotlib inline
%config InlineBackend.figure_format = 'jpeg'   # photos as JPEG, not PNG — ~10× smaller notebook
# Remove character limits for all columns of Pandas dataframes permanently in this session
pd.set_option('display.max_colwidth', None)

📦 Python  3.12.13
🔢 NumPy   2.0.2
📊 Matplotlib   3.10.0
🔥 PyTorch 2.11.0+cpu
⚠️ No GPU detected

🎲 Random seed set to 42 (Python, NumPy, PyTorch)


![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-ML-Explainers/main/assets/rainbow-divider.png)

# A neuron is embarrassingly simple

A neural network looks intimidating from the outside, but the unit it is made of fits on one line: multiply each input by a weight, add them up, add a bias.

Everything else in this course — CNNs, RNNs, transformers — is this same operation repeated with more structure. Learn it once here and the rest is arrangement.

## Compute one neuron by hand

The next cell holds three inputs, three weights and one bias. The line `(x * w).sum() + b` is the entire neuron: an elementwise multiply, a sum, and a shift.

Run it. The value it prints, z, is called the pre-activation — the neuron's raw opinion before any non-linearity touches it.

In [2]:
x = torch.tensor([0.5, -1.2, 2.0])  # 3 inputs
w = torch.tensor([0.8, 0.3, -0.5])  # 3 weights
b = torch.tensor(0.1)  # 1 bias

z = (x * w).sum() + b
# z = torch.dot(x, w) + b
print(f"Pre-activation, z = {z.item():.4f}")

Pre-activation, z = -0.8600


## The bias moves the answer alone

Weights decide how much each input matters. The bias decides where the neuron sits when every input is zero — it slides the whole output up or down.

In [3]:
BIAS = 1.0  # Increase bias by 0.9 (0.1 + 0.9 = 1.0)
b = torch.tensor(BIAS)
z = (x * w).sum() + b
print(f"Pre-activation, z = {z.item():.4f}")

Pre-activation, z = 0.0400


Now z moves by exactly +0.9 $(0.0400 - -0.8600 = 0.9)$ compared to the previous run.

## The non-linearity that makes it deep

Stack a thousand of these linear neurons and you still only get a straight line. ReLU fixes that: it keeps positive values and clamps negatives to zero.

In [4]:
pre_activations = torch.tensor([-0.9, 0.4, -0.2, 1.6, 0.0, 2.3, -1.1, 0.7])
activations = torch.relu(pre_activations)
print(f"Activations: {activations}")

Activations: tensor([0.0000, 0.4000, 0.0000, 1.6000, 0.0000, 2.3000, 0.0000, 0.7000])


![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-ML-Explainers/main/assets/rainbow-divider.png)

# References

- **[Learnism: Deep Learning — Neural Networks from Scratch Course](https://www.learnism.in/course/deep-learning)**

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-ML-Explainers/main/assets/rainbow-divider.png)